api/controllers/AuthController.js

```diff
@@ -25,49 +25,67 @@ module.exports = {
+    } else {
+      auth_data.scope = 'identity'; //Regular users only need a temporary access token with reduced scope.
     }
     passport.authenticate('reddit', auth_data)(req, res);
   },
 
   callback: function (req, res) {
     passport.authenticate('reddit', {duration: 'permanent', failureRedirect: '/login'}, async function (err, user) {
-      let login_info = JSON.parse(req.query.state);
-      if (login_info.validation !== req.session.validation) {
-        sails.warn("Failed login for /u/" + user.name + ": invalid session state");
-        return res.forbidden();
-      }
-      let finishLogin = function () {
-        req.logIn(user, function (err) {
-          if (err) {
-            sails.error('Failed login: ' + err);
-            return res.forbidden(err);
-          }
-          var url = decodeURIComponent(login_info.redirect);
-          // Don't redirect to other callback urls (this may cause infinite loops) or to absolute url paths (which might lead to other sites).
-          if (url.indexOf('/auth/reddit/callback') === 0 || /^(?:[a-z]+:)?\/\//i.test(url)) {
-            url = '/';
+      try {
+        if (err) {
+          if (err === 'banned') {
+            return res.view(403, {error: 'You have been banned from FlairHQ'});
           }
-          req.session.validation = '';
-          return res.redirect(url);
-        });
-      };
-      let modStatus = await Reddit.checkModeratorStatus(sails.config.reddit.adminRefreshToken, user.name, 'pokemontrades');
-      if (modStatus) { //User is a mod, set isMod to true
-        User.update(user.name, {isMod: true}).exec(function () {
-          /* Redirect to the mod authentication page, or to the desired url if this was mod authentication.*/
-          if (login_info.type !== 'mod') {
-            return res.redirect('/auth/reddit?loginType=mod' + (login_info.redirect ? '&redirect=' + encodeURIComponent(login_info.redirect) : ''));
-          }
-          finishLogin();
-        });
-      }
-      else if (user.isMod) { // User is not a mod, but had isMod set for some reason (e.g. maybe the user used to be a mod). Set isMod to false.
-        User.update(user.name, {isMod: false}).exec(finishLogin);
-      } else { // Regular user
-        finishLogin();
+          return res.view(403, {error: 'Sorry, something went wrong. Try logging in again.'});
+        }
+        var login_info;
+        try {
+          login_info = JSON.parse(req.query.state);
+        } catch (err) {
+          console.log('Error with parsing /u/' + user.name + '\'s session state');
+          return res.serverError(err);
+        }
+        if (login_info.validation !== req.session.validation) {
+          console.log("Failed login for /u/" + user.name + ": invalid session state");
+          return res.view(403, {error: 'You have an invalid session state. (Try logging in again.)'});
```

Comment:
Now users will be informed when they have an invalid session state or when they're banned, rather than just seeing the default 403 page.